tools/checklocks/state.go

```diff
@@ -18,23 +18,52 @@ import (
 	"fmt"
 	"go/token"
 	"go/types"
+	"sort"
 	"strings"
 	"sync/atomic"
-	"sort"
 
 	"golang.org/x/tools/go/ssa"
 )
 
 // lockType represents the current state of the lock.
 // The 0 value represents the unlocked state.
+// The valid states during analysis are:
+// unlocked, locked, deferredunlock and locked & deferredunlock
 type lockType int
 
 const (
-	unlocked lockType = iota
-	locked
-	// TODO(jamesyou): DEFERUNLOCK
+	unlocked lockType = 0
+	locked   lockType = 1 << iota
```

Comment:
The "is", "union", and "without" function are really confusing at the moment. The current representation issues can be solved with enough documentation. (Think describing an algebra with axioms.)

One options is to move away from defining locked & deferredunlock as "locked | deferred" and instead as a simple enum:
```
const (
	unlocked lockType = iota
	locked
	deferredunlock
        lockedAndDeferredUnlock
)
```
The functions and operations would need to be replaced appropriately.

An alternative is lean into the set definition. Emphasize that unlocked is an empty set. locked is equivalent to {"locked"}, etc. "is" becomes "contains" immediately. union makes sense. Without is still a bit confusing but set difference is easy to express.

Reply:
I decided to make `lockType` a simple enum instead. In particular because the state transitions should be much simpler and less confusing. The enum has clear transitions based on the current state and no longer relies on `without`.